### PR TITLE
Use the public suffix list to be less permissive on matching

### LINKED ISSUE #375: Make AutofillService domain parsing and matching more robust
Related to #372 

Currently, we are naively assuming that android app domains are of the form `x.y.z` where `y` is the domain name to be used for finding potential credentials for autofill.  Additionally, the domain matching uses a string `contains` check, which can lead to false positives and a potential for leaking credentials.

Instead, we should reverse it and use the public suffix list to tease out the top-level "private domain" based on the [PSL](https://publicsuffix.org), and domain name matching should be stricter to exact match or domain "label" descendants.

Some options:
* [Guava](https://github.com/google/guava) has [InternetDomainName](https://github.com/google/guava/wiki/InternetDomainNameExplained) to aid in the last step, though that is quite a lot for a single piece of functionality
* Fennec/FFTV/et al have [support](https://github.com/mozilla-mobile/firefox-tv/blob/master/app/src/main/java/org/mozilla/tv/firefox/utils/publicsuffix/PublicSuffixPatterns.java) that has been copy-pasta'd around (to be consolidated in [A-C #95](https://github.com/mozilla-mobile/android-components/issues/95) one day)

The preferred approach per a sidechannel conversation with @linuxwolf is that we do the latter; eventually this can be ripped out in favor of `a-c` a la `NetworkStore`, and it will keep our app size down in the long run.

### PR DESCRIPTION
**Pull request by linuxwolf**
Fixes #375

This uses `lib-publicsuffixlist` from Android-Components to tighten up the matching.

_(**Required**: this reference (one or many) will be closed upon merge. Ideally it has the acceptance criteria and designs for features or fixes related to the work in this Pull Request.)_

Connected to #???

_(Optional: other issues or pull requests related to this, but merging should not close it)_

## Testing and Review Notes

_(**Required**: steps to take to confirm this works as expected or other guidance for code, UX, and any other reviewers)_

## Screenshots or Videos

_(Optional: to clearly demonstrate the feature or fix to help with testing and reviews)_

## To Do

- add “WIP” to the PR title if pushing up but not complete nor ready for review
- [x] double check the original issue to confirm it is fully satisfied
- [x] add testing notes and screenshots in PR description to help guide reviewers
- [x] add unit tests
  - optional: consider adding instrumentation (integration/UI) tests
- consider running this branch in a debug simulator and check for memory leak notifications or any warnings
- ~request the "UX" team perform a design review (if/when applicable)~
- [x] make sure CI builds are passing (e.g.: fix lint and other errors)
- ~check on the [accessibility](https://mozilla-lockbox.github.io/lockbox-android/accessibility/) of any added UI~

### CODE REVIEW

File: app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt [hunk 1]
```diff
@@ -42,41 +55,81 @@ class LockboxAutofillService(
...
+                            .map { FillablePassword(it, serverPwd) }
+                    }
+                val zipper: (Array<Any>) -> List<FillablePassword> = {
+                    it.filter { it is FillablePassword }
+                        .map { it as FillablePassword }
+                }
+
+                when (serverPasswords.size) {
+                    0 -> Observable.just(emptyList<FillablePassword>())
+                    else -> Observable.zipIterable(
```
Comment (sashei) [inline app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt]: 🤐 ❗️

File: app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt [hunk 2]
```diff
@@ -42,41 +55,81 @@ class LockboxAutofillService(
...
+                        pslSupport.fromOrigin(serverPwd.hostname)
+                            .map { FillablePassword(it, serverPwd) }
+                    }
+                val zipper: (Array<Any>) -> List<FillablePassword> = {
+                    it.filter { it is FillablePassword }
+                        .map { it as FillablePassword }
+                }
+
+                when (serverPasswords.size) {
+                    0 -> Observable.just(emptyList<FillablePassword>())
```
Comment (sashei) [inline app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt]: nice optimization.

File: app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt [hunk 3]
```diff
@@ -42,41 +55,81 @@ class LockboxAutofillService(
...
             return
         }
 
-        if (webDomain == null) {
-            callback.onFailure("unexpected package name structure")
-            return
+        val packageName = parsedStructure.packageId ?: activityPackageName
+        val webDomain = parsedStructure.webDomain
+
+        // resolve the (webDomain || packageName) to a psl+1
```
Comment (sashei) [inline app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt]: psl+1? not quite sure what that abbreviation means.

File: app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt [hunk 4]
```diff
@@ -42,41 +55,81 @@ class LockboxAutofillService(
...
+            callback.onFailure(msg)
+        }
 
-    private fun domainFromPackage(packageName: String): String? {
-        // naively assume that the `y` from `x.y.z`-style package name is the domain
-        // untested as we will change this implementation with issue #375
-        val domainRegex = Regex("^\\w+\\.(\\w+)\\..+")
-        return domainRegex.find(packageName)?.groupValues?.get(1)
+        Observables.combineLatest(expectedDomain, passwords)
+            .subscribe(doNext, doError)
```
Comment (sashei) [inline app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt]: I personally would include the `doNext` and `doError` blocks inline with my subscription, but I think that's more of a style choice than a hard requirement.

File: app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt (hunk 1 above)
Comment (linuxwolf) [reply app/src/main/java/mozilla/lockbox/LockboxAutofillService.kt]: yeah, if there isn't at least one ... `zipIterable` barfs

**Pull request merged**